dart/constraint/ConstrainedGroup.h

```diff
@@ -107,7 +108,7 @@ class ConstrainedGroup
   std::vector<ConstraintBase*> mConstraints;
 
   ///
-  dynamics::Skeleton* mRootSkeleton;
+  std::shared_ptr<dynamics::Skeleton> mRootSkeleton;
```

Comment:
I'm still inclined to use `DyanmicsTypes.h`, and put the smart pointer types and other type definitions such as `INVALID_INDEX`.

Also, I think we could use const variable instead of unnamed enum for type safety as:
```cpp
const size_t invalidIndex = static_cast<size_t>(-1);
```
enum's default type is int unless we use enum class with name scoping like:
```cpp
enum class Index : size_t
{
  INVALID_INDEX = static_cast<size_t>(-1)
};
```